### PR TITLE
Rewrite OsuAutoGenerator

### PR DESCRIPTION
**Pull request by thomastanck**
It now uses a new algorithm that should be much easier to customise and potentially create cursor dance auto modes.

Main idea is to preprocess the beatmap into a chronological series of events before doing any actual generation.

### CODE REVIEW

**Comment by peppy** [comment]
Please don't PR from your master branch in the future (always name the branch first).

**Comment by peppy** [comment]
This algorithm currently seems to let go too early on sliders (you can notice if you set playback speed to 0.5x, there's a flash at the end of the slider).

**Comment by peppy** [comment]
There's a lot of wtf's in this code and reviewing will take quite a while. In short, though, it's very hard to follow what's going.

- I'd suggest removing all the class-level variables you added and instead passing them between steps as arguments, so it's clear as to where the flow of information is.
- I'd suggest xmldoc for methods.
- Not sure we should be using `SortedDictionary` when a `SortedList` with a comparer would be enough.
- Not sure about using `GetEnumerator`. We never do this anywhere else and it just feels weird. Maybe just use an `int index` instead?

**Comment by smoogipoo** [comment]
Since this is a very old PR, I tasked myself with getting this merged today. It's not going to happen for quite a while still, but I want to describe my thoughts and processes.

The general function of the PR looks pretty good, it's at least able to handle various 2B maps I threw at it with relative ease.

Things it couldn't do changed that story a little bit, because this PR is almost completely unreadable to me. For example, spinners are currently given a constant screen-centred position, however in the future spinners will not remain centred to the screen and take up the entire screen area.

I attempted to change this, but in the process struggled to understand how the code worked:

1. There are so many methods doing so many operations using dictionaries that it's difficulty to know what sort of data I'm looking at at any point in time.
2. All the methods to do everything are stacked into one gigantic class, and pollute the eyes when trying to narrow down the flow order between methods.
3. Some methods consider every single possible combination of hitobjects simultaneously at the same time, increasing their length and further reducing the understandability of this code.

I'm trying to resolve these issues in https://github.com/smoogipoo/osu/tree/osuautogenerator-rewrite but idk how much longer this is productive for me to keep up.

What needs to happen is the following:
* The concept of a `KeyFrame` needs to be solidified. At the moment there are 3 "hold" states that need to be followed, and two lists of points for clicks and movements.
  * In my changes, `KeyFrame` contains only actions. The "hold" state has been removed. The "click + hold" state is baked into an action of its own (`KeyFrameClickAction` with `Primary` set to false).
* The concept of intermediate frames need to be introduced. We can't have 3 different dictionaries for keyframes, "active" points, and "cursor positions".
  * I believe it should be possible for the keyframes to generate i-frames.
  * These should contain only cursor positions. They should not change the gameplay state.
* The button building step needs to take all the keyframes, generate i-frames from them, and post-process the i-frames to contain button info.
  * The result is that this step should just spit out an IEnumerable of (cursor-position, OsuAction) tuples.
  * All that `generateReplayFrames()` should be doing is taking these tuples and spitting out `OsuReplayFrame`s.

**Comment by smoogipoo** [comment]
Let's close this for now due to inactivity. While this is cool, we'll need to rethink the implementation so that there's not 1000+ lines of code inside one class.

The general concept of keyframes introduced here is something future implementations should take forward.
